### PR TITLE
Several modifications

### PR DESCRIPTION
**Pull request by sscruz**
- Adding loose to tight lepton scale factors calculated by @barbaralvarez

- Fix in the Uncertainty class so the fake rates are properly read 

- Adding JER and unclustered energy uncertainties to the signal extraction, up to the DNNs

- Fixing bug in recleaner so nFwdJet is properly defied

- Small changes to make_cards_new

- Module to find the decay of the higgs

- JEC uncertainties are properly propagated to the 2017 MET recipe (needs for https://github.com/gpetruc/nanoAOD-tools/pull/2)

### CODE REVIEW

File: TTHAnalysis/python/plotter/ttH-multilepton/mca-2lss-mcdata-frdata-splitdecays.txt [hunk 1]
```diff
@@ -6,62 +6,62 @@ incl_dataflips  : + ; IncludeMca="ttH-multilepton/mca-includes/mca-data-forFlips
 
 incl_data : + ; IncludeMca="ttH-multilepton/mca-includes/mca-data.txt"
 
-incl_sigprompt_jec_Up : + ; IncludeMca="ttH-multilepton/mca-includes/mca-2lss-sigprompt-splitdecays.txt", AddWeight="eventBTagSF_up_jes/eventBTagSF", FakeRate="ttH-multilepton/fr-jecUp.txt", SkipMe=True, PostFix="_jec_Up"
-incl_sigprompt_jec_Dn : + ; IncludeMca="ttH-multilepton/mca-includes/mca-2lss-sigprompt-splitdecays.txt", AddWeight="eventBTagSF_down_jes/eventBTagSF", FakeRate="ttH-multilepton/fr-jecDn.txt", SkipMe=True, PostFix="_jec_Dn"
+incl_sigprompt_jec_Up : + ; IncludeMca="ttH-multilepton/mca-includes/mca-2lss-sigprompt-splitdecays.txt", AddWeight="btagSF_shape_up_jes/btagSF_shape", FakeRate="ttH-multilepton/fr-jecUp.txt", SkipMe=True, PostFix="_jec_Up"
```
Comment (gpetruc) [inline TTHAnalysis/python/plotter/ttH-multilepton/mca-2lss-mcdata-frdata-splitdecays.txt:9]: in theory we shouldn't need anymore all these process variations: they were needed for the old system, bu for the new one the uncertainty file should suffice. 
That said, they may be useful to make plots of the individual shape variations

File: TTHAnalysis/python/plotter/ttH-multilepton/mca-includes/mca-3l-sigprompt-splitdecays.txt [hunk 2]
```diff
@@ -8,17 +8,17 @@ ttH_hzg+    : TTHnobb_pow : 0.5071*0.418 : LepGood1_mcMatchId!=0 && LepGood2_mcM
...
-TTZ	: TTJets_DiLepton : xsec : LepGood1_mcMatchId!=0 && LepGood2_mcMatchId!=0 && LepGood3_mcMatchId!=0 ; FillColor=ROOT.kSpring+2
-TTZ   	: TTJets_SingleLeptonFromT : xsec : LepGood1_mcMatchId!=0 && LepGood2_mcMatchId!=0 && LepGood3_mcMatchId!=0 ; FillColor=ROOT.kSpring+2
-TTZ   	: TTJets_SingleLeptonFromTbar : xsec : LepGood1_mcMatchId!=0 && LepGood2_mcMatchId!=0 && LepGood3_mcMatchId!=0 ; FillColor=ROOT.kSpring+2
+VH_hww : VHToNonbb_ll : xsec : (LepGood1_mcMatchId!=0 && LepGood2_mcMatchId!=0 && LepGood3_mcMatchId!=0)*(GenHiggsDecayMode==24) ; FillColor=ROOT.kAzure-9, years="2017\,2018"
+VH_hzz : VHToNonbb_ll : xsec : (LepGood1_mcMatchId!=0 && LepGood2_mcMatchId!=0 && LepGood3_mcMatchId!=0)*(GenHiggsDecayMode==23) ; FillColor=ROOT.kAzure-9, years="2017\,2018"
+VH_htt : VHToNonbb_ll : xsec : (LepGood1_mcMatchId!=0 && LepGood2_mcMatchId!=0 && LepGood3_mcMatchId!=0)*(GenHiggsDecayMode==15) ; FillColor=ROOT.kAzure-9, years="2017\,2018"
+
+
+TTWW	: TTWW : xsec : LepGood1_mcMatchId!=0 && LepGood2_mcMatchId!=0 && LepGood3_mcMatchId!=0 ; FillColor=ROOT.kGreen
```
Comment (gpetruc) [inline TTHAnalysis/python/plotter/ttH-multilepton/mca-includes/mca-3l-sigprompt-splitdecays.txt:21]: here I don't understand the change.
is github rendering the things differently or did you remove TTV from the 3l final state?

File: TTHAnalysis/python/plotter/ttH-multilepton/systsUnc.txt [hunk 3]
```diff
@@ -198,14 +222,28 @@ CMS_ttHl_Clos_m_bt_norm : $FRAll : .*3l_bl.* : none
...
+CMS_ttHl16_Clos_e_shape : $FRAll  : .*2lss_em.*_2016    : templateSymm ; Normalize=False, AddWeight='1.15' 
+CMS_ttHl16_Clos_m_shape : $FRAll  : .*2lss_em.*_2016    : templateSymm ; Normalize=False, AddWeight='1.15' 
+CMS_ttHl17_Clos_e_shape : $FRAll  : .*2lss_ee.*_2017    : templateSymm ; Normalize=False, AddWeight='1.3' 
+CMS_ttHl17_Clos_m_shape : $FRAll  : .*2lss_mm.*_2017    : templateSymm ; Normalize=False, AddWeight='1.3' 
+CMS_ttHl17_Clos_e_shape : $FRAll  : .*2lss_em.*_2017    : templateSymm ; Normalize=False, AddWeight='1.15' 
+CMS_ttHl17_Clos_m_shape : $FRAll  : .*2lss_em.*_2017    : templateSymm ; Normalize=False, AddWeight='1.15' 
+CMS_ttHl18_Clos_e_shape : $FRAll  : .*2lss_ee.*_2018    : templateSymm ; Normalize=False, AddWeight='1.3' 
+CMS_ttHl18_Clos_m_shape : $FRAll  : .*2lss_mm.*_2018    : templateSymm ; Normalize=False, AddWeight='1.3' 
+CMS_ttHl18_Clos_e_shape : $FRAll  : .*2lss_em.*_2018    : templateSymm ; Normalize=False, AddWeight='1.15' 
+CMS_ttHl18_Clos_m_shape : $FRAll  : .*2lss_em.*_2018    : templateSymm ; Normalize=False, AddWeight='1.15'
```
Comment (gpetruc) [inline TTHAnalysis/python/plotter/ttH-multilepton/systsUnc.txt:246]: placeholders?
the old _shape uncertainties were defined to be approximately normalized, while these are normalization only.

File: TTHAnalysis/python/tools/eventVars_2lss.py [hunk 4]
```diff
@@ -83,22 +82,28 @@ def run(self,event,Collection,metName):
...
+                met_up = getattr(event,metName+"_pt_unclustEnUp")
+                metphi_up = getattr(event,metName+"_phi_unclustEnUp")
+                met_down = getattr(event,metName+"_pt_unclustEnDown")
+                metphi_down = getattr(event,metName+"_phi_unclustEnDown")
+                if nlep > 0:
+                    allret["MT_met_lep1" + self.label + '_unclustEnUp'] = sqrt( 2*leps[0].conePt*met_up*(1-cos(leps[0].phi-metphi_up)) )
+                    allret["MT_met_lep1" + self.label + '_unclustEnDown'] = sqrt( 2*leps[0].conePt*met_down*(1-cos(leps[0].phi-metphi_down)) )
+                if nlep > 1:
+                    allret["MT_met_lep1" + self.label + '_unclustEnUp'] = sqrt( 2*leps[1].conePt*met_up*(1-cos(leps[1].phi-metphi_up)) )
+                    allret["MT_met_lep1" + self.label + '_unclustEnDown'] = sqrt( 2*leps[1].conePt*met_down*(1-cos(leps[1].phi-metphi_down)) )
```
Comment (gpetruc) [inline TTHAnalysis/python/tools/eventVars_2lss.py]: typo: lep2, not lep1 (unless I don't understand the code)

the MT_met_leplep and sum_abspx, sum_sgnpz have been intentionally removed as obsolete, or it's a mistake?

File: TTHAnalysis/python/tools/nanoAOD/higgsDecayFinder.py [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+        if not any( map(lambda x : x.pdgId == 25, genParts)):
+            self.out.fillBranch('GenHiggsDecayMode',-1)
+            return True
+        decay = None
+        for g in genParts:
+            if g.pdgId != 25: continue
+            decay = self.getHiggsDecay(genParts.index(g), genParts)
+
+        if not decay                     : self.out.fillBranch('GenHiggsDecayMode',-1); return True 
+        if 23 in decay  and 22 in decay  : self.out.fillBranch('GenHiggsDecayMode',22); return True
```
Comment (gpetruc) [inline TTHAnalysis/python/tools/nanoAOD/higgsDecayFinder.py]: I would keep H->Zg separate from H->gg (I recall fixing that in CMGTools at some point, even if apparently it didn't make it into the heppy release). e.g. 22 for `== [22,22]` and 2223 for `23 in decay  and 22 in decay`

File: TTHAnalysis/python/tools/nanoAOD/higgsDecayFinder.py [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+        elif decay == [23,23]            : self.out.fillBranch('GenHiggsDecayMode',23); return True
+        elif decay == [24,24]            : self.out.fillBranch('GenHiggsDecayMode',24); return True
+        elif decay == [15,15]            : self.out.fillBranch('GenHiggsDecayMode',15); return True 
+        elif decay == [13,13]            : self.out.fillBranch('GenHiggsDecayMode',13); return True 
+        else: 
+            self.out.fillBranch('GenHiggsDecayMode',0); return True
```
Comment (gpetruc) [inline TTHAnalysis/python/tools/nanoAOD/higgsDecayFinder.py]: now we fill with -1 for a Higgs with no decays or no Higgs, zero for not understood.
maybe better 0 for no Higgs, -1 or -99 for decay not found or not understood?

File: TTHAnalysis/python/tools/nanoAOD/lepScaleFactors.py [hunk 7]
```diff
@@ -122,19 +143,56 @@ def analyze(self, event):
...
+            if len(leps) >= 2:
+                leptonSF_2lss = self.getLooseToTight(leps[0],var,year,2) * self.getLooseToTight(leps[1],var,year,2)
+                leptonSF_2lss = leptonSF_2lss * self.getRecoToLoose(leps[0],'',year) * self.getRecoToLoose(leps[1],'',year)
+            if len(leps) >= 3:
+                leptonSF_3l   = self.getLooseToTight(leps[0],var,year,3) * self.getLooseToTight(leps[1],var,year,3) * self.getLooseToTight(leps[2],var,year,3)
+                leptonSF_3l   = leptonSF_3l *  self.getRecoToLoose(leps[0],'',year) * self.getRecoToLoose(leps[1],'',year) * self.getRecoToLoose(leps[2],'',year)
+            if len(leps) >= 4: 
+                leptonSF_4l   = self.getLooseToTight(leps[0],var,year,3) * self.getLooseToTight(leps[1],var,year,3) * self.getLooseToTight(leps[2],var,year,3) * self.getLooseToTight(leps[3],var,year,3)
+                leptonSF_4l   = leptonSF_4l *  self.getRecoToLoose(leps[0],'',year) * self.getRecoToLoose(leps[1],'',year) * self.getRecoToLoose(leps[2],'',year) * self.getRecoToLoose(leps[3],'',year)
+
```
Comment (gpetruc) [inline TTHAnalysis/python/tools/nanoAOD/lepScaleFactors.py]: Wouldn't it be cleaner to just have one loop, and have getLooseToTight and getRecoToLoose be slightlysmarter? they already need to be able to understand if it's electrons or muons and up or down.
e.g. for electrons in getRecoToLoose
`var = +1 if '_el_up' in var_str  else (-1 if '_el_dn' in var_str else 0)`
and in getLooseToTight
`var = +1 if '_el_loosetotight_up' in var_str  else (-1 if '_el_loosetotight_dn' in var_str else 0)`

**Review by gpetruc — COMMENTED** [review]
Hi,

Thanks for the changes.

I've put some comments on things that I think could be improved (but also later), or where I wanted to make sure the changes was intended and not a mistake.
There's only one that looks really suspicious to me and I would really prefer you to double-check before, on the mca-3l-sigprompt-splitdecays.txt

File: TTHAnalysis/python/plotter/ttH-multilepton/systsUnc.txt (hunk 3 above)
Comment (sscruz) [reply TTHAnalysis/python/plotter/ttH-multilepton/systsUnc.txt:246]: Yes, these are placeholders until we have the final number... Will add a "FIXME" so we don't forget

File: TTHAnalysis/python/tools/eventVars_2lss.py (hunk 4 above)
Comment (sscruz) [reply TTHAnalysis/python/tools/eventVars_2lss.py]: yes, it was a typo. regarding MT_met_lep_lep etc, they were not used anymore (I just removed commented lines)

**Comment by gpetruc** [comment]
Looks to me everything is ok now.
Do you have any other commits to push or should I merge?

**Comment by sscruz** [comment]
That's all for the moment, but I'm running the synch exercise to see that I'm not missing anything for the moment. It will take a couple of days though

**Pull request merged**